In [ ]:
from IPython.core.display import display_html
display_html("<script>Jupyter.notebook.kernel.restart()</script>",raw=True)

In [1]:
from sys import path; path.insert(1,"../src");
import fubar; from importlib import reload; reload(fubar);

***
< [GMIT Data Analytics](https://web.archive.org/web/20201029063153/https://www.gmit.ie/computer-science-and-applied-physics/higher-diploma-science-computing-data-analytics-ict) | [Home](https://github.com/SeanOhAileasa) | [README](https://github.com/SeanOhAileasa/fubar-c/blob/main/README.md) >

[![GMIT](https://github.com/SeanOhAileasa/SeanOhAileasa/blob/master/rc/gmit.png?raw=true)](https://web.archive.org/web/20201029063153/https://www.gmit.ie/computer-science-and-applied-physics/higher-diploma-science-computing-data-analytics-ict)

## GMIT - Supplemental Course Material
### Topic: [C](https://nbviewer.jupyter.org/github/SeanOhAileasa/fubar-c/blob/main/fubar-c.ipynb)
***

Supplemental course material for the ``Higher Diploma in Science in Computing (Data Analytics) (H.Dip)`` programme (75 Credits) at **Galway-Mayo Institute of Technology (GMIT)**. 

This repository is for self-study/self-examination for future modules of the H.Dip programme and preparation for future workplace opportunities.

<a id="top"></a>
***
## Table of Contents
***

## [Abstract](#abstract)

< [1-1](#abstract11)&nbsp;&nbsp;|&nbsp;&nbsp;[1-2](#abstract12)&nbsp;&nbsp;|&nbsp;&nbsp;[1-2.a](#abstract12a)&nbsp;&nbsp;|&nbsp;&nbsp;[1-3](#abstract13)&nbsp;&nbsp;|&nbsp;&nbsp;[1-4](#abstract14)&nbsp;&nbsp;|&nbsp;&nbsp;[1-4.a](#abstract14a)&nbsp;&nbsp;|&nbsp;&nbsp;[1-5](#abstract15)&nbsp;&nbsp;|&nbsp;&nbsp;[1-5.a](#abstract15a)&nbsp;&nbsp;|&nbsp;&nbsp;[1-5.b](#abstract15b)&nbsp;&nbsp;|&nbsp;&nbsp;[1-6](#abstract16)&nbsp;&nbsp;|&nbsp;&nbsp;[1-7](#abstract17)&nbsp;&nbsp;|&nbsp;&nbsp;[1-7.a](#abstract17a)&nbsp;&nbsp;|&nbsp;&nbsp;[1-7.b](#abstract17b)&nbsp;&nbsp;|&nbsp;&nbsp;[1-8](#abstract18)&nbsp;&nbsp;|&nbsp;&nbsp;[1-9](#abstract19)&nbsp;&nbsp;|&nbsp;&nbsp;[1-10](#abstract110)&nbsp;&nbsp;|&nbsp;&nbsp;[1-10.a](#abstract110a)&nbsp;&nbsp;|&nbsp;&nbsp;[1-11](#abstract111)&nbsp;&nbsp;|&nbsp;&nbsp;[1-12](#abstract112)&nbsp;&nbsp;>

<hr width=50%;>

## [Introduction](#introduction) 

- &#x1F6A7;

<hr width=50%;>

* [Statement](#statement) <br/><br/>
    * [Compound](#statementsCompound)

<hr width=50%;>

* [Data Type](#dataType) <br/><br/>
    * [Character](#dataTypesCharacter) <br/><br/>
        * [``char``](#intTypeschar) (&#8594; ``1-7.b``)&nbsp;&nbsp;|&nbsp;&nbsp;[``wchar_t``](#intTypeswchar_t) <br/><br/>
            * [Stream](#dataTypesCharacterStream) <br/><br/>
                * [``getchar``](#dataTypesCharacterStreamgetchar) (&#8594; ``1-5.b`` &#8594; ``1-6``) <br/><br/>
                * [``putchar``](#dataTypesCharacterStreamputchar) (&#8594; ``1-5.b``) <br/><br/>
    * [Integer](#dataTypesIntegers) <br/><br/>     
        * [``int``](#intTypesint) <br/><br/>          
        * [Qualifier](#intTypesQualifier) <br/><br/>
            * [``short``](#intTypesshort) <br/><br/>
            * [``long``](#intTypeslong) <br/><br/>
            * [Signed/Unsigned](#intSignedversusUnsigned) <br/><br/>
    * [Single-Precision Floating Point](#dataTypesSingle-PrecisionFloatingPoint) <br/><br/>
        * [``float``](#dataTypesDoublePrecisionFloatingPointfloat) <br/><br/>
    * [Double-Precision Floating Point](#dataTypesDoublePrecisionFloatingPoint) <br/><br/>
        * [``double``](#dataTypesDoublePrecisionFloatingPointdouble) &#x1F6A7; <br/><br/>
        * [Qualifier](#dataTypesDoubleQualifier) &#x1F6A7; <br/><br/>
            * [``long``](#dataTypesDoubleQualifierslong) <br/><br/>
    * [Symbolic Constant](#dataTypeSymbolicConstant) (&#8594; ``1.5``) <br/><br/>
        * [``#define``](#dataTypeSymbolicConstantshashdefine) (&#8594; ``1.5.a``)

<hr width=50%;>

* [Declaration](#declaration)

<hr width=50%;>

* [Operators and Expressions](#operatorsexpressions) &#x1F6A7; <br/><br/>
    * [Increment](#operatorsIncrementDecrement)&nbsp;&nbsp;|&nbsp;&nbsp;[Decrement](#operatorsIncrementDecrement) (&#8594; ``1-7.a``) <br/><br/>
    * [Assignment](#operatorsIncrementDecrementAssignment) (&#8594; ``1.10.a``)     

<hr width=50%;>

* [Control Flow](#controlFlow) <br/><br/>
    * [``while``](#controlFlowwhile) (&#8594; ``1-7.a``) <br/><br/>
    * [``for``](#controlFlowfor) (&#8594; ``1-7.a``)

<hr width=50%;>

* [Function](#function) <br/><br/>
    * [``main``](#functionsmain)

<hr width=50%;>

* [Preprocessor](#preprocessor) <br/><br/>
    * [Directive](#preprocessorDirective) <br/><br/>
        * [``include``](#preprocessorDirectivehashinclude) <br/><br/>
            * [Header File](#preprocessorDirectivehashincludeHeaderFile) <br/><br/>
                * [``stdio.h``](#preprocessorDirectivehashincludeHeaderFilesstdioh) <br/><br/>
                    * [``EOF``](#preprocessorDirectivehashincludeHeaderFilesstdiohsymbolicconstantEOF) (&#8594; ``1-5.b`` &#8594; ``1-6`` &#8594; ``1-7``) <br/><br/>
                    * [``printf``](#preprocessorDirectivehashincludeHeaderFilesstdiohprintf) <br/><br/>
                    * [``scanf``](#preprocessorDirectivehashincludeHeaderFilesstdiohscanf) <br/><br/>
                * [``limits.h``](#preprocessorDirectivehashincludeHeaderFileslimitsh) <br/><br/>
                * [``float.h``](#preprocessorDirectivehashincludeHeaderFilesfloath)

<hr width=50%;>

## [Solution](#solution)

- &#x1F6A7;

<hr width=50%;>

## [Conclusion](#conclusion)

- &#x1F6A7;

<hr width=50%;>

***
## END

< [Table of Contents](#top) | [Introduction](#introduction) | [Solution](#solution) | [Conclusion](#conclusion) | [References](#references) >
<a id="abstract"></a>
***
## Abstract
***

< [Table of Contents](#top) | [References](#references) >
<a id="abstract11"></a>
***
###### 1-1
***

In [1]:
%%writefile hw.c
    /* Exercise 1-1.
Input:
Process: (printf)
Output:
*/
#include <stdio.h>
int main()
{
    printf("hw\n");
}

Writing hw.c


In [2]:
%%script bash
cc hw.c

In [3]:
%%script bash
./a.out

hw


In [4]:
%%script bash
rm -rf hw.c a.out

< [Table of Contents](#top) | [References](#references) >
<a id="abstract12"></a>
***
###### 1-2
***

In [1]:
%%writefile hw.c
    /* Exercise 1-2: What happens when printf argument string contains \c.
Input:
Process: (printf)
Output:
*/
#include <stdio.h>
int main()
{
    printf("hw\c");
}

Writing hw.c


In [2]:
%%script bash
cc hw.c

hw.c: In function 'main':
hw.c:9:12: warning: unknown escape sequence: '\c'
    9 |     printf("hw\c");
      |            ^~~~~~


In [3]:
%%script bash
rm -rf hw.c a.out

< [Table of Contents](#top) | [References](#references) >
<a id="abstract12a"></a>
***
###### 1-2.a
***

In [1]:
import ROOT

Welcome to JupyROOT 6.24/00


- computation begins with assignment statements - setting the variables to initial values [2]

In [2]:
%%cpp
int nCelsius,nLower=0,nUpper=300,nStep=20,nFahr=nLower;

- using the formula $C=(5{\div}9)(F-32)$ to create the table of fahrenheit to celsius by first dividing $5{\div}9$ would result in integer division truncation (fractional part is discarded) - since $5$ and $9$ are integers then $5{\div}9$ would be truncated to zero and so all the celsius temperatures would be reported as zero [2] 

In [3]:
%%cpp
5/9

(int) 0


In [4]:
%%cpp
5*(nFahr-32)/9

(int) -17


- using integer arithmetic results in the celsius temperatures not being very accurate ($0^{\circ}F$ is actually about $-17.8^{\circ}C$) - declaring ``nFahr`` and ``nCelsius`` to be ``float`` allows the formula for conversion to be written in a more natural way (unable to use $5{\div}9$ given integer division would truncate it to zero) - a decimal point in a constant indicates that it is floating point so $5.0{\div}9.0$ is not truncated because it is the ratio of two floating-point values [2]

In [5]:
%%cpp
5.0/9.0

(double) 0.55555556


- if an arithmetic operator has integer operands then an integer operation is performed - if an arithmetic operator has one floating-point operand and one integer operand then the integer will be converted to floating point before the operation is done [2]

In [6]:
%%cpp
5.0/9

(double) 0.55555556


In [7]:
%%cpp
5/9.0

(double) 0.55555556


In [8]:
%%cpp
float nFahr,nCelsius;

- writing ``nFahr-32`` then the ``32`` would be automatically converted to floating point - writing floating-point constants with explicit decimal points even when they are integral values emphasises their floating-point nature for human readers [2]

In [9]:
%%cpp
nFahr-32

(float) -32.0000f


- the assignment ``nFahr=nLower`` and the test ``while (nFahr<=nUpper)`` also works in the natural way (the ``int`` is converted to ``float`` before the operation is done) [2]

In [10]:
%%cpp
nFahr

(float) 0.00000f


In [11]:
%%cpp
nLower

(int) 0


In [12]:
%%cpp
nUpper

(int) 300


- each line of the table is computed the same way using a loop that repeats once per output line [2]

In [13]:
%%cpp
while (nFahr<=nUpper)
{ 
    nCelsius=(5.0/9)*(nFahr-32);
    printf("%3.0f%6.1f\n",nFahr,nCelsius);
    nFahr=nFahr+nStep;
}

  0 -17.8
 20  -6.7
 40   4.4
 60  15.6
 80  26.7
100  37.8
120  48.9
140  60.0
160  71.1
180  82.2
200  93.3
220 104.4
240 115.6
260 126.7
280 137.8
300 148.9


< [Table of Contents](#top) | [References](#references) >
<a id="abstract13"></a>
***
###### 1-3
***

In [1]:
%%writefile fc.c
    /* Exercise 1-3: Using formula C=(5/9)(F-32) to create table of fahrenheit to celsius.
Input:
Process: (printf)
Output:
*/
#include <stdio.h>
int main()
{
    int nLower=0,nUpper=300,nStep=20;
    float nFahr=nLower,nCelsius;
    printf("Fahrenheit - Celsius\n");
    while (nFahr<=nUpper)
    {
        nCelsius=(5.0/9)*(nFahr-32);
        printf("%6.0f%13.1f\n",nFahr,nCelsius);
        nFahr=nFahr+nStep;
    }
}

Writing fc.c


In [2]:
%%script bash
cc fc.c

In [3]:
%%script bash
./a.out

Fahrenheit - Celsius
     0        -17.8
    20         -6.7
    40          4.4
    60         15.6
    80         26.7
   100         37.8
   120         48.9
   140         60.0
   160         71.1
   180         82.2
   200         93.3
   220        104.4
   240        115.6
   260        126.7
   280        137.8
   300        148.9


In [4]:
%%script bash
rm -rf fc.c a.out

< [Table of Contents](#top) | [References](#references) >
<a id="abstract14"></a>
***
###### 1-4
***

In [1]:
%%writefile cf.c
    /* Exercise 1-4: Repurpose formula C=(5/9)(F-32) to create table of celsius to fahrenheit.
Input:
Process: (printf)
Output:
*/
#include <stdio.h>
int main()
{
    int nLower=0,nUpper=300,nStep=20;
    float nFahr,nCelsius=nLower;
    printf("Celsius - Fahrenheit\n");
    while (nCelsius<=nUpper)
    {
        nFahr=nCelsius/(5.0/9)+32;
        printf("%4.0f%13.1f\n",nCelsius,nFahr);
        nCelsius=nCelsius+nStep;
    }
}

Writing cf.c


In [2]:
%%script bash
cc cf.c

In [3]:
%%script bash
./a.out

Celsius - Fahrenheit
   0         32.0
  20         68.0
  40        104.0
  60        140.0
  80        176.0
 100        212.0
 120        248.0
 140        284.0
 160        320.0
 180        356.0
 200        392.0
 220        428.0
 240        464.0
 260        500.0
 280        536.0
 300        572.0


In [4]:
%%script bash
rm -rf cf.c a.out

< [Table of Contents](#top) | [References](#references) >
<a id="abstract14a"></a>
***
###### 1-4.a
***

- variation on the temperature converter (refer ``1-2.a``) [2]

In [1]:
import ROOT

Welcome to JupyROOT 6.24/00


- eliminate most of the variables with ``nFahr`` remaining (make it an ``int``) [2]

In [2]:
%%cpp
int nFahr;

- variables ``nLower`` and ``nUpper`` and ``nStep`` appear only as constants in the ``for`` statement [2]

- the expression that computes the ``nCelsius`` now appears as the third argument of the ``printf`` instead of as a separate assignment statement [2]

In [3]:
%%cpp
for (nFahr=0; nFahr<=300; nFahr=nFahr+20)
    printf("%3d %6.1f\n",nFahr,(5.0/9)*(nFahr-32));

  0  -17.8
 20   -6.7
 40    4.4
 60   15.6
 80   26.7
100   37.8
120   48.9
140   60.0
160   71.1
180   82.2
200   93.3
220  104.4
240  115.6
260  126.7
280  137.8
300  148.9


< [Table of Contents](#top) | [References](#references) >
<a id="abstract15"></a>
***
###### 1-5
***

In [1]:
%%writefile cf.c
    /* Exercise 1-5: Using the formula C=(5/9)(F-32) create table of celsius to fahrenheit in reverse order.
Input:
Process: (printf)
Output:
*/
#include <stdio.h>
int main()
{
    int nFahr;
    for (nFahr=300; nFahr>=0; nFahr-=20)
        printf("%3d %6.1f\n",nFahr,(5.0/9)*(nFahr-32));
}

Writing cf.c


In [2]:
%%script bash
cc cf.c

In [3]:
%%script bash
./a.out

300  148.9
280  137.8
260  126.7
240  115.6
220  104.4
200   93.3
180   82.2
160   71.1
140   60.0
120   48.9
100   37.8
 80   26.7
 60   15.6
 40    4.4
 20   -6.7
  0  -17.8


In [4]:
%%script bash
rm -rf cf.c a.out

< [Table of Contents](#top) | [References](#references) >
<a id="abstract15a"></a>
***
###### 1-5.a
***

In [1]:
import ROOT

Welcome to JupyROOT 6.24/00


- the quantities ``LOWER`` and ``UPPER`` and ``STEP`` are symbolic constants [2]

- these are not variables so do not appear in declarations and are conventionally written in upper case so can be readily distinguished from lower case variable names [2]

- note there is no ``;`` at the end [2]

In [2]:
%%cpp
#define LOWER 0 // table lower limit
#define UPPER 300 //  table upper limit
#define STEP 20 // table step size

In [3]:
%%cpp
int nFahr;

In [4]:
%%cpp
for (nFahr=UPPER; nFahr>=LOWER; nFahr-=STEP)
    printf("%3d%6.1f\n",nFahr,(5.0/9)*(nFahr-32));

300 148.9
280 137.8
260 126.7
240 115.6
220 104.4
200  93.3
180  82.2
160  71.1
140  60.0
120  48.9
100  37.8
 80  26.7
 60  15.6
 40   4.4
 20  -6.7
  0 -17.8


< [Table of Contents](#top) | [References](#references) >
<a id="abstract15b"></a>
***
###### 1-5.b
***

- can write a surprising amount of useful code without knowing anything more about input and output using ``getchar`` and ``putchar`` [2]

- a character on the keyboard or screen like everything else is stored internally just as a bit pattern - type ``char`` is specifically meant for storing such character data but any type ``int`` can be used (``int`` is used for a subtle but important reason) [2]

- the problem is distinguishing the end of the input from valid data - ``getchar`` returns a distinctive value when there is no more input (a value that cannot be confused with any real character - ``EOF``) [2]

- variable ``c`` must be declared a type big enough to hold any value that ``getchar`` returns and big enough to hold ``EOF`` therefore using ``int`` [2] - on linux and OS X the character to input to cause an ``EOF`` is ``Ctrl-D`` and for windows its ``Ctrl-Z`` [[stackoverflow.com](https://stackoverflow.com/questions/1782080/what-is-eof-in-the-c-programming-language)]

- ``EOF`` is an integer defined in ``stdio.h`` but the specific number value does not matter as long as it is not the same as any ``char`` value - using the symbolic constant assures that nothing in the program depends on the specific numeric value [2]

In [1]:
%%writefile getchar_putchar.c
    /* Version 1: A program that copies its input to its output one character at a time.
Input: 
Process: (getchar; putchar)
    read a character
    while (character is not end-of-file indicator)
        output the character just read
        read a character
Output: copy input to output
*/
#include <stdio.h>
int main()
{
    int c;
    c=getchar();
    while (c!=EOF)
    {
        putchar(c);
        c=getchar();
    }
}

Writing getchar_putchar.c


In [2]:
%%script bash
cc getchar_putchar.c
# ./a.out

In [3]:
%%script bash
rm -rf getchar_putchar.c ./a.out

- any assignment is an expression and has a value which is the value of the LHS after the assignment - an assignment can appear as part of a larger expression [2]

- the assignment of a character to variable ``c`` can be put inside the test part of a ``while`` - the ``while`` gets a character and assigns it to ``c`` then tests whether the character was the ``EOF`` [2]

In [1]:
%%writefile getchar_putchar.c
    /* Version 2: A program that copies its input to its output one character at a time.
Input: 
Process: centralising the input with only one reference to getchar 
    (getchar; putchar)
Output: shrinks the program (more compact) and once the idiom is mastered is easier to read
*/
#include <stdio.h>
int main()
{
    int c;
    while ((c=getchar())!=EOF) // != higher precedence
    {
        putchar(c);
    }
}

Writing getchar_putchar.c


In [2]:
%%script bash
cc getchar_putchar.c
# ./a.out

In [3]:
%%script bash
rm -rf getchar_putchar.c ./a.out

< [Table of Contents](#top) | [References](#references) >
<a id="abstract16"></a>
***
###### 1-6
***

In [1]:
%%writefile getcharEOF.c
    /* Exercise 1-6: Verify that the expression getchar() != EOF is 0 or 1.
Input: 
Process: format specifier %d is a signed decimal integer (%i specifies integer) 
    (printf; getchar)
Output: 1 (true) if character and 0 (false) if EOF
*/
#include <stdio.h>
int main()
{
    printf("%i\n",getchar()!=EOF);
}

Writing getcharEOF.c


In [2]:
%%script bash
cc getcharEOF.c
# ./a.out

In [3]:
%%script bash
rm -rf getcharEOF.c a.out

In [4]:
import ROOT

Welcome to JupyROOT 6.24/00


In [5]:
%%cpp
getchar()!=EOF // 0 specifies false

(bool) false


< [Table of Contents](#top) | [References](#references) >
<a id="abstract17"></a>
***
###### 1-7
***

In [1]:
%%writefile EOF.c
    /* Exercise 1-7: Program to print the value of EOF.
Input: 
Process: (printf)
Output: 
*/
#include <stdio.h>
int main()
{
    printf("%i\n",EOF);
}

Writing EOF.c


In [2]:
%%script bash
cc EOF.c
# ./a.out

In [3]:
%%script bash
rm -rf EOF.c a.out

In [4]:
import ROOT

Welcome to JupyROOT 6.24/00


In [5]:
%%cpp
EOF

(int) -1


< [Table of Contents](#top) | [References](#references) >
<a id="abstract17a"></a>
***
###### 1-7.a
***

- accumulating count in variable ``n`` of type ``long`` instead of ``int`` (conversion specification ``%ld`` tells ``printf`` that the corresponding argument is a ``long`` integer [2]

- note if the input contains no characters then the ``while`` test fails on the very first call to ``getchar`` and the program produces $0$ (the correct answer) [2]

In [1]:
%%writefile countCharacters.c
    /* Version 1: Program to count characters in input.
Input: 
Process: (getchar; printf)
Output: counts characters (including enter key)
*/
#include <stdio.h>
int main()
{
    long n=0;
    while (getchar()!=EOF)
        ++n;
    printf("%ld\n",n);
}

Writing countCharacters.c


In [2]:
%%script bash
cc countCharacters.c
# ./a.out

In [3]:
%%script bash
rm -rf countCharacters.c

- to cope with bigger numbers using a ``double`` precision ``float`` and using a ``for`` statement instead of a ``while`` [2]

- conversion specification ``%f`` tells ``printf`` that the corresponding argument is a ``float`` and ``double`` - using ``%.0f`` suppresses printing of the decimal point and the fraction part (zero) [2]

- the body of the ``for`` loop is empty because all the work is done in the test and increment part - the grammatical rules of ``C`` require that a ``for`` statement has a body (the isolated semicolon is called a ``null statement`` to satisfy this requirement) [2] 

- note if the input contains no characters then the ``for`` test fails on the very first call to ``getchar`` and the program produces $0$ (the correct answer) [2]

In [1]:
%%writefile bigCount.c
    /* Version 2: Program to count characters in input.
Input: 
Process: (getchar; printf)
Output:
*/
#include <stdio.h>
int main()
{
    double n;
    for (n=0;getchar()!=EOF;++n)
        ; // null statement
    printf("%.0f\n",n);
}

Writing bigCount.c


In [2]:
%%script bash
cc bigCount.c
# ./a.out

In [3]:
%%script bash
rm -rf bigCount.c a.out

< [Table of Contents](#top) | [References](#references) >
<a id="abstract17b"></a>
***
###### 1-7.b
***

- ``'\n'`` (single quotes) is the value of the newline character which is $10$ in ASCII and is a single character (in expressions is just an integer) [2]

- ``"\n"`` is a string constant that happens to contain only one character [2]

In [1]:
%%writefile countingLines.c
    /* Program to count input lines.
Input: 
Process: (getchar; printf)
Output:
*/
#include <stdio.h>
int main()
{
    int c,nL=0;
    while ((c=getchar())!=EOF)
        if (c=='\n') // "\n" string constant
            ++nL;
    printf("%d\n",nL);
}

Writing countingLines.c


In [2]:
%%script bash
cc countingLines.c
# ./a.out

In [3]:
%%script bash
rm -rf countingLines.c a.out

< [Table of Contents](#top) | [References](#references) >
<a id="abstract18"></a>
***
###### 1-8
***

In [1]:
%%writefile countBlankTabNewline.c
    /* Exercise 1-8: Write a program to count blanks, tabs, and newlines.
Input: 
Process: (getchar; printf)
Output:
*/
#include <stdio.h>
int main()
{
    int c,nB=0,nT=0,nL=0;
    while ((c=getchar())!=EOF)
    {
        if (c==' ')
            ++nB;
        else if (c=='\t')
            ++nT;
        else if (c=='\n')
            ++nL;
    }
    printf("Blanks = %d\nTab = %d\nNewline = %d\n",nB,nT,nL);    
}

Writing countBlankTabNewline.c


In [2]:
%%script bash
cc countBlankTabNewline.c
# ./a.out

In [3]:
%%script bash
rm -rf countBlankTabNewline.c a.out

< [Table of Contents](#top) | [References](#references) >
<a id="abstract19"></a>
***
###### 1-9
***

In [1]:
%%writefile input2OutputBlanks.c
    /* Exercise 1-9:  Write a program to copy its input to its output,
        replacing each string of one or more blanks by a single blank.
Input:
Process: while "current" not end-of-file (getchar); 
    if "current" or "previous" not blank (i. ASCII decimal or; ii. character constant); 
        display "current" (putchar);
    capture "current" in "previous"
Output:
*/
#include <stdio.h>
int main()
{
    int c,p; // "current" "previous" empty
    while ((c=getchar())!=EOF)
    {
        if (c!=32||p!=' ')
            putchar(c);
        p=c;
    }
}

Writing input2OutputBlanks.c


In [2]:
%%script bash
cc input2OutputBlanks.c
# ./a.out

In [3]:
%%script bash
rm -rf input2OutputBlanks.c a.out

< [Table of Contents](#top) | [References](#references) >
<a id="abstract110"></a>
***
###### 1-10
***

In [1]:
%%writefile replaceTabBackspaceBackslash.c
    /* Write a program to copy its input to its output, replacing each
    tab by \t, each backspace by \b, and each backslash by \\. This makes
    tabs and backspaces visible in an unambiguous way.
Input:
Process: (getchar; putchar)
Output:
*/
#include <stdio.h>
int main()
{
    int c;
    while((c=getchar())!=EOF)
    {
        if (c=='\t')
        {
            putchar('\\');
            c='t';
        }
        else if (c=='\b')
        {
            putchar('\\');
            c='b';
        }
        else if (c=='\\')
        {
            putchar('\\');
            c='\\';
        }
        putchar(c);
    }
}

Writing replaceTabBackspaceBackslash.c


In [2]:
%%script bash
cc replaceTabBackspaceBackslash.c
# ./a.out

In [3]:
%%script bash
rm -rf replaceTabBackspaceBackslash.c a.out

< [Table of Contents](#top) | [References](#references) >
<a id="abstract110a"></a>
***
###### 1-10.a
***

- program that counts lines and words and characters with the loose definition that a word is any sequence of characters that does not contain a blank or tab or newline [2]

- every time the program encounters the first character of a word it counts one more word [2]

In [1]:
import ROOT

Welcome to JupyROOT 6.24/00


- using the symbolic constants ``IN`` and ``OUT`` instead of the literal values $1$ and $0$ (makes the program more readable) - this makes it easier to make extensive changes in programs where magic numbers appear only as symbolic constants [2] 

In [2]:
%%cpp
#define IN 1 // inside a word
#define OUT 0 // outside a word

- variable ``state`` records whether the program is currently in a word or not (initially not a word) [2]

In [3]:
%%cpp
int state=OUT;

In [4]:
%%cpp
int c,nl,nw,nc;
nl=nw=nc=0;

- ``while`` gets a character and assigns it to ``c`` then tests whether the character is ``EOF`` [2]

- ``if`` the variable ``c`` is blank ``' '`` there is no need to test whether it is a ``'\n'`` or ``'\t'`` so these tests are not made (not particularly important but is significant in more complicated situations) [2]

In [1]:
%%writefile wc.c
    /* A program that counts lines, words, and characters, with the loose
    definition that a word is any sequence of characters that does not
    contain a blank, tab or newline.
Input:
Process: (getchar)
Output: count lines and words and characters in input
*/
#include <stdio.h>
#define IN 1 // inside a word
#define OUT 0 // outside a word
int main()
{
    int state=OUT;
    int c,nl,nw,nc;
    nl=nw=nc=0;
    while ((c=getchar())!=EOF)
    {
        ++nc; // count number character
        if (c=='\n')
            ++nl; // count number line
        if (c==' '||c=='\n'||c=='\t')
            state=OUT;
        else if (state==OUT)
        {
            state=IN;
            ++nw;
        }
    }
    printf("lines: %d\nwords:%d\ncharacters:%d\n",nl,nw,nc);
}

Writing wc.c


In [2]:
%%script bash
cc wc.c
# ./a.out

< [Table of Contents](#top) | [References](#references) >
<a id="abstract111"></a>
***
###### 1-11
***

- test ``1.10.a`` (word count program) - input most likely to uncover bugs include inputs such as hyphenated words or words with apostrophise etc (counted as one word) given the definition of a word as characters grouped together [2]

In [3]:
%%script bash
cc wc.c
# ./a.out

In [4]:
%%script bash
rm -rf wc.c a.out

< [Table of Contents](#top) | [References](#references) >
<a id="abstract112"></a>
***
###### 1-12
***

In [1]:
%%writefile printInputOneWordPerLine.c
    /* Write a program that prints its input one word per line.
Input:
Process: (getchar; putchar)
Output:
*/
#include <stdio.h>
#define IN 1
#define OUT 0
int main()
{
    int c;
    int state=OUT; // outside a word
    while ((c=getchar())!= EOF)
    {
        if (c==' '||c=='\t') 
        {
            if (state==IN)
                putchar('\n');
            state=OUT;
        } 
        else
        {
            putchar(c);
            state=IN;            
        }
    }
}

Writing printInputOneWordPerLine.c


In [2]:
%%script bash
cc printInputOneWordPerLine.c

In [3]:
%%script bash
rm -rf printInputOneWordPerLine.c a.out

***
## END

< [Table of Contents](#top) | [Abstract](#abstract) | [Solution](#solution) | [Conclusion](#conclusion) | [References](#references) >
<a id="introduction"></a>
***
## Introduction
***

&#x1F6A7;

< [Table of Contents](#top) | [References](#references) >
<a id="statement"></a>
***
### Statement
***

- statements can contains one or more expressions - an expression being a function call or other operator(s) (mathematical/bitwise etc) - can be combined in a single statement although its important to use good judgement and not make things too complex (hard to read) [1]

- its important from a readability point of view to put statements on a single line although sometimes that does not work (sometimes statements are to long and span multiple lines) - ``C`` does not care where statements are put [1]

- the fact that statements on are a line by themselves so that they could be readable means nothing to the compiler - can write a whole program on one line albeit dumb thing to do - the point is that ``C`` is a free format syntax language whereby a line means nothing - Fortran on the otherhand is a little different whereby a line is very important and even column numbers are important [1]

< [Table of Contents](#top) | [References](#references) >
<a id="statementsCompound"></a>
***
###### Compound
***

- every statement must end in a semicolon - this differs from some languages like Pascal where a semicolon is a statement separator meaning there is no need to have the last statement followed by a semicolon [1]

- an exception is when a group of statements act almost as if it were a single statement - the ``if`` statement to make some decisions based on a condition being ``true`` then executing some statements - basically groups a number of statements together as a single unit called a compound statement - need to make a block out of it - a block being something that is surrounded with curly braces [1]

- compound statements group single statements (end with semicolons) - the block of statements are surrounded by curly braces [1]

***
### END

< [Table of Contents](#top) | [References](#references) >
<a id="dataType"></a>
***
### Data Type
***

- basic data types in ``C`` include ``char`` / ``int`` / ``float`` / ``double`` [2]

< [Table of Contents](#top) | [References](#references) >
<a id="dataTypesCharacter"></a>
***
###### Character
***

- a single byte capable of holding one character in the local character set [2]

< [Table of Contents](#top) | [References](#references) >
<a id="intTypeschar"></a>
***
###### ``char``
***

- special form of integer dealing with characters in a program - an underlying character set allows the encoding of each character to some number (the most commonly used is ASCII) [1]

- lowercase $a$ encoded decimal $97$ - characters are really just numbers albeit treated differently in programs - how they are stored depends on the encoding used - ASCII being the most common (other types of encoding exist) [1]

- a character written between single quotes represents an integer value equal to the numerical value of the character in the machines character set (called a character constant) [2]

- ``'A'`` is a character constant and in the ASCII character set is $65$ (internal presentation of the character ``A``) [2]

In [1]:
import ROOT

Welcome to JupyROOT 6.24/00


In [2]:
%%cpp
int c='A';
c

(int) 65


In [3]:
%%cpp
printf("%c = %d\n",c,c);

A = 65


- EBCDIC (Extended Binary Coded Decimal Interchange Code) is considerably different than ASCII [1]

- a standard which is a superset of ASCII called ISO 8859 includes some characters that are needed in other countries besides the United States [1]

- Unicode (universal encoding) is a 16-bit encoding as opposed to 8-bit (ASCII / EBCDIC / ISO 8859) - purpose of Unicode is to allow a single encoding for languages throughout the modern business world - cannot just use ASCII and go to Japan (entirely different alphabet) - the same is true throughout the world (many different encodings) - to write programs that can adapt themselves to different encodings is extremely difficult and has been one of the biggest impediments to programs being made for the international market - Unicode is an attempt to solve the problem and it turns out that Unicode is the encoding used in ``Java`` [1]

In [4]:
%%cpp
CHAR_MIN

(int) -128


In [5]:
%%cpp
CHAR_MAX

(int) 127


< [Table of Contents](#top) | [References](#references) >
<a id="intTypeswchar_t"></a>
***
###### ``wchar_t``
***

- no encoding is specified for ``C`` or ``C++`` (left open) [1]

- for ``C`` and ``C++`` the basic character data type is ``char`` (byte) - thats pretty much all there is to it - think of a byte and thats what a ``char`` (narrow character) is - the basic ``C`` character does not allow the use of Unicode and for this reason have defined a new type which is available in both ``C`` and ``C++`` called ``wchar_t`` (wide character) [1]

- some functions in the library deal with white characters - in most systems a white character is 16-bits and so it allows the use of Unicode but all of this needs to be supported by the compiler and by the platform [1]

< [Table of Contents](#top) | [References](#references) >
<a id="dataTypesCharacterStream"></a>
***
###### Stream
***

- the model of input and output supported by the standard library is very simple - text input or output regardless of where it originates or where it goes to is dealt with as a streams of characters [2]

- a ``text stream`` is a sequence of characters divided into lines - each line consists of zero or more characters followed by a newline character [2]

- it is the responsibility of the library to make each input or output stream conform to this model - using the library there is no need to worry about how lines are represented outside the program [2]

- the standard library provides several functions for reading or writing one character at a time of which ``getchar`` and ``putchar`` are the simplest [2]

< [Table of Contents](#top) | [References](#references) >
<a id="dataTypesCharacterStreamgetchar"></a>
***
###### ``getchar``
***

- reads the next input character from a text stream and returns that as its value [2]

- variable ``c`` contains the next character of input [2]

```c
c=getchar();
```

< [Table of Contents](#top) | [References](#references) >
<a id="dataTypesCharacterStreamputchar"></a>
***
###### ``putchar``
***

- prints a character each time it is called - prints the contents of the integer variable ``c`` as a character usually on the screen [2]

```c
putchar(c);
```

- calls to ``putchar`` and ``printf`` may be interleaved (the output appear in the order in which the calls are made) [2]

< [Table of Contents](#top) | [References](#references) >
<a id="dataTypesIntegers"></a>
***
###### Integer
***

- fundamental data type and basic building blocks of all entities in a program [1]

- integers come in different sizes in ``C`` which should not be surprising since its a systems programming language - the different sizes are for different purposes [1]

- matches the word size of the machine - word refers to the number of bits processed by a computers CPU in one go (typically 64-bits) [1]

< [Table of Contents](#top) | [References](#references) >
<a id="intTypesint"></a>
***
###### ``int``
***

- typically reflects the natural size of integers on the host machine [2]

In [1]:
import ROOT

Welcome to JupyROOT 6.24/00


In [2]:
%%cpp
INT_MIN

(int) -2147483648


In [3]:
%%cpp
INT_MAX

(int) 2147483647


< [Table of Contents](#top) | [References](#references) >
<a id="intTypesQualifier"></a>
***
###### Qualifier
***

- there are a number of qualifiers that can be applied to the basic data types - the integer qualifiers are ``short`` and ``long`` with the word ``int`` typically omitted in such declarations [2]

- on most platforms ``short`` and ``int`` and ``long`` are really not all distinct - will never see a platform where all three are distinct although possible [1]

- the intent is that ``short`` and ``long`` should provide different lengths of integers where practical - ``int`` will normally be the natural size for a particular machine - ``short`` is often 16-bits long and ``int`` either 16 or 32-bits - each compiler is free to choose appropriate sizes for its own hardware subject only to the restriction that ``short`` and ``int`` are at least 16-bits and ``long`` is at least 32-bits and ``short`` is no longer than ``int`` which is no longer than ``long`` [2]

< [Table of Contents](#top) | [References](#references) >
<a id="intTypesshort"></a>
***
###### ``short``
***

In [1]:
import ROOT

Welcome to JupyROOT 6.24/00


In [2]:
%%cpp
SHRT_MIN

(int) -32768


In [3]:
%%cpp
SHRT_MAX

(int) 32767


< [Table of Contents](#top) | [References](#references) >
<a id="intTypeslong"></a>
***
###### ``long``
***

- ``long`` integers are at least 32 bits although on some machines ``int`` and ``long`` are the same size [2]

In [1]:
import ROOT

Welcome to JupyROOT 6.24/00


In [2]:
%%cpp
LONG_MIN

(long) -9223372036854775808


In [3]:
%%cpp
LONG_MAX

(long) 9223372036854775807


< [Table of Contents](#top) | [References](#references) >
<a id="intSignedversusUnsigned"></a>
***
###### Signed/Unsigned
***

- assuming most machines are 32-bit or bigger then this is the size that an integer would be [1]

- the ``C`` standard guarantees that an integer will be at least 16-bits - compilers under ``DOS`` are 16-bits which means it can store a number up to $32767$ - reasonable to think that 16-bits would be $2^{16-1}$ which happens to be $65535$ but it turns out that integers can be ``signed`` and ``unsigned`` (by default ``signed``) [1]

- ``Java`` only has ``signed`` integers [1]

In [1]:
2**16-1

65535

- qualifiers may be applied to ``char`` or any integer - unsigned numbers are always positive or zero and obey the laws of arithmetic modulo $2^n$ (where n is the number of data type bits) - if ``char`` is 8-bits then an ``unsigned char`` has values between $0$ and $255$ while ``signed char`` has values between $-128$ and $127$ (in a twos complement machine) - whether plain ``chars`` are ``signed`` or ``unsigned`` is machine-dependent but printable characters are always positive [2]

In [2]:
2**8 # include zero

256

< [Table of Contents](#top) | [References](#references) >
<a id="dataTypesSingle-PrecisionFloatingPoint"></a>
***
###### Single-Precision Floating Point
***

- as with integers the size of floating-point objects are implementation-defined - ``float`` and ``double`` and ``long double`` could represent one or two or three distinct sizes [2]

< [Table of Contents](#top) | [References](#references) >
<a id="dataTypesDoublePrecisionFloatingPointfloat"></a>
***
###### ``float``
***

- a ``float`` is typically a 32-bit quantity with at least six significant digits and has a magnitude generally between about $10^{-38}$ and $10^{+38}$ [2]

< [Table of Contents](#top) | [References](#references) >
<a id="dataTypesDoublePrecisionFloatingPoint"></a>
***
###### Double-Precision Floating Point
***

- as with integers the size of floating-point objects are implementation-defined - ``float`` and ``double`` and ``long double`` could represent one or two or three distinct sizes [2]

< [Table of Contents](#top) | [References](#references) >
<a id="dataTypesDoublePrecisionFloatingPointdouble"></a>
***
###### ``double``
***

&#x1F6A7;

< [Table of Contents](#top) | [References](#references) >
<a id="dataTypesDoubleQualifier"></a>
***
###### Qualifier
***

&#x1F6A7;

< [Table of Contents](#top) | [References](#references) >
<a id="dataTypesDoubleQualifierslong"></a>
***
###### ``long``
***

- as with integers the size of floating-point objects are implementation-defined - ``float`` and ``double`` and ``long double`` could represent one or two or three distinct sizes - type ``long double`` specifies extended-precision floating point [2]

< [Table of Contents](#top) | [References](#references) >
<a id="dataTypeSymbolicConstant"></a>
***
###### Symbolic Constant
***

- its bad practice to bury ``magic numbers`` in a program - these numbers convey little information to someone who might have to read the program later and are hard to change in a systematic way [2]

- one way to deal with ``magic numbers`` is to give them meaningful names [2]

< [Table of Contents](#top) | [References](#references) >
<a id="dataTypeSymbolicConstantshashdefine"></a>
***
###### ``#define``
***

- defines a symbolic name or symbolic constant to be a particular string of characters [2]

- any occurence of ``name`` (not in quotes and not part of another name) will be replaced by the corresponding ``replacement text`` [2]

- the ``name`` has the same form as a variable name (a sequence of letters and digits that begins with a letter) - the ``replacement text`` can be any sequence of characters (not limited to numbers) [2]

```c
#define name replacement text
```

***
### END

< [Table of Contents](#top) | [References](#references) >
<a id="declaration"></a>
***
### Declaration
***

- all variables must be declared before use and usually at the beginning of a function before any executable statements - a declaration announces the properties of variables consisting of a type name and a list of variables [2]

***
### END

< [Table of Contents](#top) | [References](#references) >
<a id="operatorsexpressions"></a>
***
### Operators and Expressions
***

&#x1F6A7;

< [Table of Contents](#top) | [References](#references) >
<a id="operatorsIncrementDecrement"></a>
***
###### Increment&nbsp;&nbsp;|&nbsp;&nbsp;Decrement
***

- operator ``++`` increments by one - could write ``n=n+1`` but ``++n`` is more concise and often more efficient (corresponding operator ``--`` to decrement by one) [2]

- can be prefix or postfix operators - the two forms have different values in expressions albeit both ``++n`` and ``n++`` both increment ``n`` [2]

< [Table of Contents](#top) | [References](#references) >
<a id="operatorsIncrementDecrementAssignment"></a>
***
###### Assignment
***

- an assignment is an expression with a value [2]

- assignments associate from right to left [2]

In [1]:
import ROOT

Welcome to JupyROOT 6.24/00


In [2]:
%%cpp
int x,y,z;
x=(y=(z=1));

In [3]:
%%cpp
printf("%d - %d - %d\n",x,y,z);

1 - 1 - 1

In [4]:
%%cpp
x=y=z=0;

In [5]:
%%cpp
x

(int) 0


In [6]:
%%cpp
y

(int) 0


In [7]:
%%cpp
z

(int) 0


***
### END

< [Table of Contents](#top) | [References](#references) >
<a id="controlFlow"></a>
***
### Control Flow
***

- the choice between the ``while`` and ``for`` is arbitrary and based on which seems clearer [2]

- the ``for`` is usually appropriate for loops in which the initialisation and increment are single statements and logically related since it is more compact than ``while`` and it keeps the loop control statements together in one place [2]

< [Table of Contents](#top) | [References](#references) >
<a id="controlFlowwhile"></a>
***
###### ``while``
***

- the condition in parentheses is tested - if ``true`` the body of the loop is executed - the condition is re-tested - if ``true`` the body is executed again - when the test becomes ``false`` then the loop ends and execution continues at the statement that follows the loop [2]

- the body of a ``while`` can be one or more statements enclosed in braces or a single statement without braces [2]

In [1]:
import ROOT

Welcome to JupyROOT 6.24/00


In [2]:
%%cpp
int i=1,j=5;
while (i<j)
    i=2*i;

In [3]:
%%cpp
i

(int) 8


< [Table of Contents](#top) | [References](#references) >
<a id="controlFlowfor"></a>
***
###### ``for``
***

- a generalisation of the ``while`` [2]

- the first part is the initialisation and done once before the loop proper is entered [2]

- the second part is the test or condition that controls the loop [2]

- the third part is the increment step with the second part (condition) re-evaluated [2]

***
### END

< [Table of Contents](#top) | [References](#references) >
<a id="function"></a>
***
### Function
***

- functions contain statements that specify the computing operations to be done [2]

< [Table of Contents](#top) | [References](#references) >
<a id="functionsmain"></a>
***
###### ``main``
***

- the ``C`` program begins executing at the beginning of ``main`` which means every program must have a ``main`` somewhere [2]

***
### END

< [Table of Contents](#top) | [References](#references) >
<a id="preprocessor"></a>
***
### Preprocessor
***

- a ``C`` program is a collection of functions (being the equivalent of a procedure or subroutine - itself a collection of statements enclosed by braces) in addition to some optional global variables/constants [1]

- having all of the functions of a very large ``C`` program in a single source file could be awkward - can modularise a program so as to having many files which after compilation and linking comprise the program [1]

- two source files (``s1.c`` and ``s2.c``) at the second level from the top [1]

- each source file includes a couple of header files (other files that have some code) - the source file ``s1.c`` includes ``f1.h`` and ``f2.h`` - the source file ``s2.c`` includes ``f2.h`` and ``f3.h`` [1]

- imagining the depiction of this executable ``C`` program is a project - it has a couple of main modules and is divided up into a couple of files - each of these files is including things that are defined elsewhere - in particular ``f2.h`` is included in both places so do not want to have to type that separately in the two source files so its put into its own file [1]

- the types of things that go into header files are mainly just function declarations which are functions without a body - a function declaration identifies the function (name / return type / any other information that is necessary) - type once and reuse as necessary [1]

- the fundamental compilation unit (the thing that gets compiled from source code into machine code) is the combination of the source file and its included text from any files that it chooses to include [1]

- compiling ``s1.c`` and ``s2.c`` the compiler will automatically open and read the information in the header files included - the entire result gets put out in an object file (``s1.obj`` and ``s2.obj`` respectively) - is platform dependent (using Windows ``.obj`` - using Unix ``.o``) [1]

- once everything is compiled then its linked together and made ready to execute with the final result being an executable file ``prog.exe`` [1]

![Preprocessor-Executable](https://github.com/SeanOhAileasa/SeanOhAileasa/blob/master/rc/fubar-c/executable.png?raw=true)

***(Image) C. Allison, "Thinking in C," [MindView LLC](https://web.archive.org/web/20210617100433/http://ix.cs.uoregon.edu/~norris/cis330books/ThinkingInC/Index.html), (n.d.).***

- idea of including files into a compilation stream shows an important step in the ``C`` compilation process (before the program is translated from ``C`` source code into machine code) - the preprocessing phase of the compiler will first process all of the preprocessing operations [1]

< [Table of Contents](#top) | [References](#references) >
<a id="preprocessorDirective"></a>
***
###### Directive
***

&#x1F6A7;

< [Table of Contents](#top) | [References](#references) >
<a id="preprocessorDirectivehashinclude"></a>
***
###### ``include``
***

- includes another file into the compilation stream - the pound sign ``#`` plus ``include`` is a special directive in C [1]

- it invokes a preprocessor allowing the inclusion of other text into the program without retyping - these including declarations of special functions that are in the standard ``C`` library and in particular the header ``stdio.h`` (usually call ``standard IO.h``) [1]

- the include directive occurs before the actual translation process - putting function declarations and maybe some constant declarations into a header file - its different in ``C++`` given there is other things you might want to put in there but for ``C`` that is pretty much it [1]

< [Table of Contents](#top) | [References](#references) >
<a id="preprocessorDirectivehashincludeHeaderFile"></a>
***
###### Header File
***

- the standard ``C`` library makes use of the preprocessor - in other words each ``C`` compiler will make available header files such as ``stdio.h`` [1]

- there is quite a number of header files in the standard ``C`` library and its possible to distinguish between standard header files from any other header files - other header files might be created by someone else in an organisation [1]

- the standard header file uses angle brackets (``<`` ``>``) to delimit the standard header name - to the compiler any header which is delimited by angle brackets will occur in a particular place in the file system (in some directory) - there exists a standard system directory where those headers go and during compilation that is where the compiler will look [1]

- to create user-defined headers use double quotes (``"`` ``"``) instead of the angle brackets (``<`` ``>``) - the compiler will look in the current directory for the file - there are some rules for finding them in other places but those are platform dependent - Unix and Windows can have environment variables - some compilers will look at a particular environment variable which has a set of path names (different directories to search) in case it does not find what its looking for in the current directory [1]

< [Table of Contents](#top) | [References](#references) >
<a id="preprocessorDirectivehashincludeHeaderFilesstdioh"></a>
***
###### ``stdio.h``
***

- there are certain channels of input and output that are assumed (streams) [1]

- the functions available in the standard ``C`` library (in particular ``stdio.h``) provide I/O services for using what is called the console (basically the screen) and the keyboard or for explicitly doing file input and output [1]

- three basic categories of types of I/O: i) console I/O (goes to standard input and output); ii) file I/O (explicit file input and output - naming a file for reading from or writing to); iii) in-core or in-memory formatting [1]

- there is three streams that are predefined in C:

1. standard in standard input (``stdin``): the console functions in the standard ``C`` library will always read from standard input - standard input is by default the keyboard - there are ways to change this called redirection - just keep in mind that standard input is where a number of functions in the standard ``C`` library get their input [1]


2. standard out (``stdout``): analogous situation for output (hooked to the screen) - functions like ``puts`` send output to standard output [1]


3. standard error (``stderr``): channel or stream for output which is also connected to the screen by default - it is logically different than standard output - can send things separately to standard output and standard error - if not redirecting them to files or other devices then will see all on the screen together but it does allow for a logically separate place to put them and if its chosen to have standard output not go to the screen then can do that [1]

< [Table of Contents](#top) | [References](#references) >
<a id="preprocessorDirectivehashincludeHeaderFilesstdiohsymbolicconstantEOF"></a>
***
###### ``EOF``
***

- end of file symbolic constant [2] 

In [1]:
import ROOT

Welcome to JupyROOT 6.24/00


In [2]:
%%cpp
EOF

(int) -1


< [Table of Contents](#top) | [References](#references) >
<a id="preprocessorDirectivehashincludeHeaderFilesstdiohprintf"></a>
***
###### ``printf``
***

- general-purpose output formatting function [2] ``printf`` works similar to function ``scanf`` [1]

- the function is not part of the ``C`` language as there is no input or output defined in ``C`` itself - it is a useful function from the standard library of functions that are normally accessible to ``C`` programs - the behaviour of ``printf`` is defined in the ANSI standard however its properties should be the same with any compiler and library that conforms to the standard [2] 

- the first argument is a format specification and any literal text will just be echoed to standard out - any conversion specification ``%`` included within the string of characters to be printed indicates arguments to be substituted and in what form it is to be printed (``%d`` specifies an integer argument) [2] 

In [1]:
import ROOT

Welcome to JupyROOT 6.24/00


- the conversion specification ``%3.0f`` indicates a floating-point number to be printed with at least three characters wide with no decimal point and no fraction digits [2]

In [2]:
%%cpp
float n=2.0/3;
printf("%3.0f\n",n);

  1


- the conversion specification ``%6.1f`` prints at least size characters wide with 1 digit after the decimal point [2]

In [3]:
%%cpp
printf("%6.1f\n",n);

   0.7


- width and precision may be omitted from a specification [2]

In [4]:
%%cpp
printf("%6d\n...print as decimal number - at least 6 characters wide\n",int(n));

     0
...print as decimal number - at least 6 characters wide


In [5]:
%%cpp
printf("%f\n...print as floating point\n",n);

0.666667
...print as floating point


In [6]:
%%cpp
printf("%9f\n...print as floating point - at least 9 characters wide\n",n);

 0.666667
...print as floating point - at least 9 characters wide


In [7]:
%%cpp
printf("%.2f\n...print as floating point - 2 characters after decimal point\n",n);

0.67
...print as floating point - 2 characters after decimal point


- among other the function ``printf`` also recognises ``%o`` for octal and ``%x`` for hexadecimal and ``%c`` for character and ``%s`` for character string and ``%%`` for $\%$ itself [2]

< [Table of Contents](#top) | [References](#references) >
<a id="preprocessorDirectivehashincludeHeaderFilesstdiohscanf"></a>
***
###### ``scanf``
***

- ``scanf`` is like ``printf`` except that it reads input instead of writing output [2]

- scans standard input for the input to read - the first argument is the format string (integer ``%d``) - the special character specifies that the character that follows is a code which describes the type of input - reading a decimal integer (base-10) by calling function ``scanf`` then standard input will be scanned for numbers - when it runs out (perhaps the user hits enter or space or something else) then the number that results from those numerals typed in will be stored [1]

- take it as fact an ampersand ``&`` is required in front of whatever variable is read with ``scanf`` (relates to pointers) [1]

In [1]:
import ROOT

Welcome to JupyROOT 6.24/00


In [2]:
%%writefile sf.c
#include <stdio.h>
int main()
{
    int n;
    scanf("%d",&n);
    printf("%d",n);
}

Writing sf.c


In [3]:
%%script bash
cc sf.c

In [4]:
%%script bash
./a.out # terminal input integer

0

In [5]:
%%script bash
rm -rf sf.c a.out

< [Table of Contents](#top) | [References](#references) >
<a id="preprocessorDirectivehashincludeHeaderFileslimitsh"></a>
***
###### ``limits.h``
***

- its a little disconcerting (frustrating) that its not possible to specify what the range of the numeric types are (different for different systems) [1]

- can be troublesome to write programs that will be run on different systems (want portability) - ``C`` is supposed to be a portable language and to some degree it is but the fact is it is a systems programming language (more system friendly than it is programmer friendly) - while its not true to a large degree it certainly is true with the underlying numeric types - cannot specify a universal value for the largest ``int`` or the smallest ``int`` as it differs from platform to platform [1]

- its impossible to find out by querying it in a program by examining the symbolic constants for all of these sizes along with other properties of the machine and compiler [1][2]

- integer limits are defined in ``limits.h`` - including this header file or looking up the documentation and examining the names of those constants then its possible can find out what is the largest ``int`` or the largest ``short`` or the largest ``long`` or the smallest as well etc - this does give some programmatic control over the numeric properties of the system - the same thing applies to the floating point numbers as well and this information can be found in the header ``float.h`` [1]

In [1]:
%%script bash
pwd
# cat ../../../usr/include/limits.h

- for integral limits can access constants (typically written with capital letters - convention thats been around for a long time) [13]

< [Table of Contents](#top) | [References](#references) >
<a id="preprocessorDirectivehashincludeHeaderFilesfloath"></a>
***
###### ``float.h``
***

- contain symbolic constants for all floating-point sizes along with other properties of the machine and compiler [2]

***
### END

***
## END

< [Table of Contents](#top) | [Abstract](#abstract) | [Introduction](#introduction) | [Conclusion](#conclusion) | [References](#references) >
<a id="solution"></a>
***
## Solution
***

&#x1F6A7;

<!-- - refer [Appendix](#appendix-df) to view the full dataset <a id="go-back-df"></a> -->
<!-- print(fGetModuleType(___x___),fDocumentation(___x___)) -->
***
## END

< [Table of Contents](#top) | [Abstract](#abstract) | [Introduction](#introduction) | [Solution](#solution) | [References](#references) >
<a id="conclusion"></a>
***
## Conclusion
***

&#x1F6A7;

***
## END

< [Table of Contents](#top) | [Abstract](#abstract) | [Introduction](#introduction) | [Solution](#solution) | [Conclusion](#conclusion) | [References](#references) >
<a id="references"></a>
***
## References
***

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[1] C. Allison, "Thinking in C," [MindView LLC](https://web.archive.org/web/20210617100433/http://ix.cs.uoregon.edu/~norris/cis330books/ThinkingInC/Index.html), (n.d.).

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[2] B. W. Kernighan, D. M. Ritchie, "The C Programming Language (2nd Edition)," Prentice Hall, 1988.

***
## END

< [Table of Contents](#top) | [Abstract](#abstract) | [Introduction](#introduction) | [Solution](#solution) | [Conclusion](#conclusion) | [References](#references) >
<a id="appendix"></a>
***
## Appendix
***

###### Images
***

![gumball](https://github.com/SeanOhAileasa/SeanOhAileasa/blob/master/rc/cta/analysing/gumball.png?raw=true)

***(Image) P. Mannion, "03 Analysing Algorithms Part 1," GMIT, October 2017.***

###### Tables
***

|aaa|bbbbbb|ccccccccc|
|:------------------------- | :------------------------- | :------------------------- |
|aaa|bbbbbb|ccccccccc|

###### References
***

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[1] x. xxx, "xxx," xxx, JanXXXry 2XX1.


&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[0] Anaconda (built-in), "range," Docstring, Conda:4.9.2 (Python:3.8.5).

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[0] Anaconda (plot), "matplotlib.pyplot," Docstring, Conda:4.9.2 (Python:3.8.5).

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[0] D. Carr, "CTA - Lecture 1 Part 1," GMIT, January 2021.

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[0] A. Beatty, "xxx," GMIT, January 202x.

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[0] M. R. Fellows, I. Parberry, "SIGACT Trying to Get Children Excited About CS," [ianparberry](https://web.archive.org/web/20210220071000/http://ianparberry.com/research/cseducation), January 1993.

###### fubar-000

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[0] A. Beatty, "prog 0.0 introduction (Module: Programming for Cybersecurity)," GMIT (Programme: Certificate in Cybersecurity Operations), January 2021.

***
## END

# END JUPYTER NOTEBOOK